server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeShutdownAllocationDecider.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.decider;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.NodesShutdownMetadata;
+import org.elasticsearch.cluster.metadata.SingleNodeShutdownMetadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.Nullable;
+
+/**
+ * An allocation decider that prevents shards from being allocated to a
+ * node that is in the process of shutting down.
+ *
+ * In short: No shards can be allocated to, or remain on, a node which is
+ * shutting down for removal. Primary shards cannot be allocated to, or remain
+ * on, a node which is shutting down for restart.
+ */
+public class NodeShutdownAllocationDecider extends AllocationDecider {
+    private static final Logger logger = LogManager.getLogger(NodeShutdownAllocationDecider.class);
+
+    private static final String NAME = "node_shutdown";
+
+    /**
+     * Determines if a shard can be allocated to a particular node, based on whether that node is shutting down or not.
+     */
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        final SingleNodeShutdownMetadata thisNodeShutdownMetadata = getNodeShutdownMetadata(allocation.metadata(), node.nodeId());
+
+        if (thisNodeShutdownMetadata == null) {
+            // There's no shutdown metadata for this node, return yes.
+            return allocation.decision(Decision.YES, NAME, "this node is not currently shutting down");
+        }
+
+        switch (thisNodeShutdownMetadata.getType()) {
+            case REMOVE:
+                return allocation.decision(Decision.NO, NAME, "node [%s] is preparing to be removed from the cluster", node.nodeId());
+            case RESTART:
+                return allocation.decision(
+                    Decision.YES,
+                    NAME,
+                    "node [%s] is preparing to restart, but will remain in the cluster",
+                    node.nodeId()
+                );
+            default:
+                logger.error(
```

Comment:
If we ever get into this state, I think this is going to spam the logs like crazy and be unsilenceable (because it's at the error level). I mean, hopefully we don't get into this state, but I think I'd prefer to just have the assert fail our builds. What do you think?

Reply:
You're right, I didn't think about how spammy it would be. I do think that a way to identify that we're hitting this condition in the field besides looking for a `YES` decision with no explanation is good (what can I say, I'm paranoid), but I'll drop it down to `DEBUG` so at least it won't spam by default. Sound good?